### PR TITLE
PropertyManager::asJson should not serialize properties that are not enabled

### LINKED ISSUE #21321: PropertyManager::asJson should not serialize properties that are not enabled
`PropertyManager::asJson` will serialize a property `A` depending on the truthiness of other property `B` independent of the value of `B`. This leads to confusion when `B` is `false`. 
In the picture, property `NormRunNumbers` has a value that will not be used. However, the string representation of the algorithm becomes:

```
{"name":"BASISReduction","properties":{"DoIndividual":"1","EnergyBins":"-120,0.4,120","MaskFile":"/SNS/BSS/shared/autoreduce/new_masks_08_12_2015/BASIS_Mask_default_111.xml",
"NoMonitorNorm":"1","NormRunNumbers":"76185-76201","RunNumbers":"75778-75803"},"version":1}
```
Anyone reading this string may think that `NormRunNumbers` was in fact used

screenshot

### PR DESCRIPTION
**Pull request by jmborr**
- Included a condition to check if property is enabled
- unit test

**To test:**
Run the following algorithm declaration in MantidPlot's python window:
```
from __future__ import print_function
class Hello(PythonAlgorithm):
    def __init__(self):
        PythonAlgorithm.__init__(self)
    def name(self):
        return 'Hello'
    def category(self):
        return "General"
    def PyInit(self):
        self.declareProperty('Greet', True, direction=Direction.Input,
                             doc='Do I personally greet you?')
        ifGreet = EnabledWhenProperty('Greet', PropertyCriterion.IsDefault)
        self.declareProperty('Name', '', 'Enter a name')
        self.setPropertySettings('Name', ifGreet)
    def PyExec(self):
        print(self)
AlgorithmFactory.subscribe(Hello)
```
Now execute `Hello` from MantidPlot with some value for `Name`, first with `Greet` option enabled  and then with `Greet` disabled. The  output shown in the log widget for the first case should be:

```
{"name":"Hello","properties":{"Name":"Horace"},"version":1}
```

(I chose "Horace" for `Name`), and for the second case:

```
{"name":"Hello","properties":{"Greet":"0"},"version":1
```

"Horace" is not printed because of disabled `Greet`.

Deserialization:

```
import json
for json_str in ('{"name":"Hello","properties":{"Name": "Horace"},"version":1}',
                 '{"name":"Hello","properties":{"Greet":"0"},"version":1}'):
    x = json.loads(json_str)
    alg = AlgorithmManager.create('Hello')
    alg.initialize()
    for name,value in x['properties'].items():
        alg.setProperty(str(name), value)
    alg.execute()
```

Should result in the following output:

```
{"name":"Hello","properties":{"Name":"Horace"},"version":1}

{"name":"Hello","properties":{"Greet":"0"},"version":1}
```

Fixes #21321

[Release Notes ](https://github.com/mantidproject/mantid/blob/a9c794d35de2a5966c8b37876ea1216d653f83a1/docs/source/release/v3.12.0/framework.rst#core-functionality)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by peterfpeterson — APPROVED** [review]
It works as advertised.

File: Framework/Kernel/src/PropertyManager.cpp [hunk 1]
```diff
@@ -500,7 +501,12 @@ ::Json::Value PropertyManager::asJson(bool withDefaultValues) const {
   const size_t count = propertyCount();
   for (size_t i = 0; i < count; ++i) {
     Property *p = getPointerToPropertyOrdinal(static_cast<int>(i));
```
Comment (quantumsteve) [inline Framework/Kernel/src/PropertyManager.cpp]: (linting issue) Given that `getPointerToPropertyOrdinal` accepts an `int`, shouldn't the loop variable type be `int` and `count` be cast outside of the loop?

**Review by quantumsteve — CHANGES_REQUESTED** [review]
Needs a little cleanup before being merged.

**Review by quantumsteve — APPROVED** [review]

**Comment by jmborr** [comment]
@peterfpeterson @quantumsteve :zap: thanks! :zap:

**Pull request merged**